### PR TITLE
Add custom emote command responses

### PR DESCRIPTION
**Pull request by alazymeme**
Due to the increase in bttv emote slots, I've added a requested feature that allows users to replace the default !bttv/ffz/subemotes command response with their own.

Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable
- [x] Documentation in docs/ or install-docs/ was updated, if applicable

Fixes https://github.com/pajbot/pajbot/issues/910

### CODE REVIEW

**Comment by alazymeme** [comment]
Thanks to @pajlada for his help on this image

File: pajbot/modules/basic/emotes.py [hunk 1]
```diff
@@ -79,8 +114,14 @@ def print_twitch_emotes(self, **rest):
...
-        for message in messages:
-            self.bot.say(message)
+        if self.settings["custom_sub_response"] != "":
+            custom_message = self.settings["custom_sub_response"]
+
+        if custom_message == "":
+            for message in messages:
+                self.bot.say(message)
+        else:
+            self.bot.say(custom_message.format(streamer=streamer, source=source))
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:124]:
```suggestion
        if self.settings["custom_sub_response"] != "":
            streamer = self.bot.streamer_display
            custom_message = self.settings["custom_sub_response"]
            self.bot.say(custom_message.format(streamer=streamer, source=source))
        else:
            manager = self.bot.emote_manager.twitch_emote_manager

            messages = split_into_chunks_with_prefix(
                [
                    {"prefix": "Subscriber emotes:", "parts": [e.code for e in manager.tier_one_emotes]},
                    {"prefix": "T2:", "parts": [e.code for e in manager.tier_two_emotes]},
                    {"prefix": "T3:", "parts": [e.code for e in manager.tier_three_emotes]},
                ],
                default=f"Looks like {StreamHelper.get_streamer()} has no subscriber emotes! :(",
            )

            for message in messages:
                self.bot.say(message)            
```

File: pajbot/modules/basic/emotes.py [hunk 2]
```diff
@@ -105,8 +146,8 @@ def do_reload(bot, source, **rest):
         )
 
     def print_cmd(self, manager, examples):
-        def do_print(**rest):
-            self.print_emotes(manager)
+        def do_print(bot, source, **rest):
+            self.print_emotes(source, manager)
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:150]:
```suggestion
        def do_print(source, **rest):
            self.print_emotes(source, manager)
```

File: pajbot/modules/basic/emotes.py [hunk 3]
```diff
@@ -50,26 +50,61 @@ class EmotesModule(BaseModule):
         ModuleSetting(
             key="enable_subemotes", label="Enable !subemotes command", type="boolean", required=True, default=True
         ),
+        ModuleSetting(
+            key="custom_sub_response",
+            label="Enable a custom response to the !subemotes command. Leave empty to disable the message. | Available arguments: {source}, {streamer}",
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:55]:
```suggestion
            label="A custom message to override the default !subemotes output format. Leave empty to use default format (1 or multiple messages showing all emotes). | Available arguments: {source}, {streamer}",
```

File: pajbot/modules/basic/emotes.py [hunk 4]
```diff
@@ -50,26 +50,61 @@ class EmotesModule(BaseModule):
...
+            placeholder="@{source}, Channel sub emotes can be found here: https://twitchemotes.com/channels/11148817",
+            default="",
+            constraints={"max_str_len": 400},
+        ),
         ModuleSetting(
             key="enable_ffzemotes", label="Enable !ffzemotes command", type="boolean", required=True, default=True
         ),
+        ModuleSetting(
+            key="custom_ffz_response",
+            label="Enable a custom response to the !ffzemotes command. Leave empty to disable the message. | Available arguments: {source}, {streamer}",
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:67]:
```suggestion
            label="A custom message to override the default !ffzemotes output format. Leave empty to use default format (1 or multiple messages showing all emotes). | Available arguments: {source}, {streamer}",
```

File: pajbot/modules/basic/emotes.py [hunk 5]
```diff
@@ -50,26 +50,61 @@ class EmotesModule(BaseModule):
...
+            placeholder="@{source}, Channel FFZ emotes can be found here: https://www.frankerfacez.com/channel/pajlada",
+            default="",
+            constraints={"max_str_len": 400},
+        ),
         ModuleSetting(
             key="enable_bttvemotes", label="Enable !bttvemotes command", type="boolean", required=True, default=True
         ),
+        ModuleSetting(
+            key="custom_bttv_response",
+            label="Enable a custom response to the !bttvemotes command. Leave empty to disable the message. | Available arguments: {source}, {streamer}",
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:79]:
```suggestion
            label="A custom message to override the default !bttvemotes output format. Leave empty to use default format (1 or multiple messages showing all emotes). | Available arguments: {source}, {streamer}",
```

File: pajbot/modules/basic/emotes.py [hunk 6]
```diff
@@ -50,26 +50,61 @@ class EmotesModule(BaseModule):
...
-            self.bot.say(message)
+        if self.settings[f"custom_{manager.friendly_name.lower()}_response"] != "":
+            custom_message = self.settings[f"custom_{manager.friendly_name.lower()}_response"]
 
-    def print_twitch_emotes(self, **rest):
+        if custom_message == "":
+            for message in messages:
+                self.bot.say(message)
+        else:
+            self.bot.say(custom_message.format(streamer=streamer, source=source))
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:103]:
```suggestion
        if self.settings[f"custom_{manager.friendly_name.lower()}_response"] != "":
            streamer = self.bot.streamer_display
            custom_message = self.settings[f"custom_{manager.friendly_name.lower()}_response"]
            self.bot.say(custom_message.format(streamer=streamer, source=source))
        else:
            emotes = manager.channel_emotes
            messages = split_into_chunks_with_prefix(
                [{"prefix": f"{manager.friendly_name} emotes:", "parts": [e.code for e in emotes]}],
                default=f"No {manager.friendly_name} Emotes active in this chat :(",
            )

            for message in messages:
                self.bot.say(message)
```

**Comment by alazymeme** [comment]
> for the suggested changes, you might need to reformat afterwards, I made these in the web editor
> 
> Additionally, I cant place review comments there, but you need to update the command examples as well: https://github.com/pajbot/pajbot/pull/940/files#diff-96ac7c3796efc215492133288ccc3c3eR161-R162 (three times).

I semi-fixed the variables in the command examples. Only thing I'm having issues with is importing the `streamer_display` variable.

File: pajbot/modules/basic/emotes.py [hunk 7]
```diff
@@ -144,7 +144,7 @@ def do_print(source, **rest):
             self.print_emotes(source, manager)
 
         if self.settings[f"custom_{manager.friendly_name.lower()}_response"]:
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:146]:
```suggestion
        if self.settings[f"custom_{manager.friendly_name.lower()}_response"] != "":
```

File: pajbot/modules/basic/emotes.py [hunk 8]
```diff
@@ -165,7 +165,7 @@ def do_print(source, **rest):
 
     def print_twitch_cmd(self):
         if self.settings["custom_sub_response"]:
```
Comment (RAnders00) [inline pajbot/modules/basic/emotes.py:167]:
```suggestion
        if self.settings["custom_sub_response"] != "":
```

File: pajbot/modules/basic/emotes.py (hunk 7 above)
Comment (alazymeme) [reply pajbot/modules/basic/emotes.py:146]: The idea of this is to specifically match nothing. Is this still applicable?

Comment (alazymeme) [reply pajbot/modules/basic/emotes.py:146]: nevermind, I misunderstoon

**Comment by RAnders00** [comment]
I don't like the use of `StreamHelper` because it will be the streamer's login name and not display name. What exactly about the old approach did not work?

**Comment by alazymeme** [comment]
> I don't like the use of `StreamHelper` because it will be the streamer's login name and not display name. What exactly about the old approach did not work?

`AttributeError: 'NoneType' object has no attribute 'self.streamer'` (I believe this was the error).

No matter what I did, this error seemed to have followed...

**Comment by alazymeme** [comment]
@pajlada Here's the error I get. The error occurs for the web interface and the page ends up erroring out: `AttributeError: 'NoneType' object has no attribute 'streamer_display'`

image

**Comment by pajlada** [comment]
You can keep using StreamHelper for now (which will result in us using lowercase usernames, just like before). We should inject the streamers display name into StreamHelper so it can be used in situations like this where the modules command output and examples are generated in the same spot and we try to use the custom response in the example.
#945

**Comment by alazymeme** [comment]
> You can keep using StreamHelper for now (which will result in us using lowercase usernames, just like before). We should inject the streamers display name into StreamHelper so it can be used in situations like this where the modules command output and examples are generated in the same spot and we try to use the custom response in the example.

Considering this was the only issue, this pr can be merged, no? Or is something else yet to be done on it?

**Pull request merged**